---
title: "Motif discovery in selected peaks"
author: "Katharina Hembach"
date: "9/6/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## FASTA sequences of selected peaks

For HOMER, we need fasta files of the peaks sequences and corresponding background sequences.

We use the predicted peak regions as they. If that does not give us good peaks we can try to center a window of 41nts around the peak center (region with the highest coverage).

```{r}
library(BSgenome.Mmusculus.UCSC.mm10)
library(GenomicAlignments)
library(stringr)
library(ggplot2)
library(here)
```

```{r peak-import}
genome <- BSgenome.Mmusculus.UCSC.mm10

sns <- import(here("analysis/deduplicated/MA_plot_selection/selected_peaks/selected_peaks_SNS_pValue_1e-05.bed"))
hom <- import(here("analysis/deduplicated/MA_plot_selection/selected_peaks/selected_peaks_total_cortex_pValue_3e-03.bed"))
```

We annotate each peak to exons, introns, 3'UTR and 5'UTR.
```{r peak-annotation, cache = TRUE}
source(here("R", "data_import.R"))

gtf_file <- "/home/Shared/data/annotation/Mouse/Ensembl_GRCm38.90/gtf/Mus_musculus.GRCm38.90.gtf"
gtf <- import(gtf_file)
anno <- prepare_anno(gtf)
anno[["gene"]] <- NULL

sns_anno <- lapply(anno, function(x) {
  res <- subsetByOverlaps(sns, x, type = "any")
  seqlevelsStyle(res) <- "UCSC"
  res})

hom_anno <- lapply(anno, function(x) {
  res <- subsetByOverlaps(hom, x, type = "any")
  seqlevelsStyle(res) <- "UCSC"
  res})
names(hom_anno) <- names(anno)
```

Write the peak fasta files:
```{r get-peak-seqs, cache = TRUE}
seqlevelsStyle(sns) <- "UCSC"
seqlevelsStyle(hom) <- "UCSC"
sns_seq <- getSeq(genome, sns)
hom_seq <- getSeq(genome, hom)
export(sns_seq, con = here("analysis/deduplicated/MA_plot_selection/peaks_fasta/top_peaks_SNS_pValue_1e-05.fasta"), format = "fasta")
export(hom_seq, con = here("analysis/deduplicated/MA_plot_selection/peaks_fasta/top_peaks_total_cortex_pValue_3e-03.fasta"), format = "fasta")
```


## Background sequences

We generate background sets for both the SNS and total cortex samples. 

```{r generate-bg-function}
#' Generate background sequences
#'
#' Function to simulate background peaks with the median peak length in a 
#' specific annotation type, not overlapping with any predicted peaks.
#'
#' @param anno_type 
#' @param anno 
#' @param peaks 
#' @param coverage 
#' @param n_bg_seqs 
#' @param median_peak_length 
#' @param bg_widths 
#' @param type 
#' @param genome_seq 
#' @param prefix 
#' @param suffix 
#' @param outdir 
#'
#' @return
#' @export
#'
#' @examples
generate_bg_seq <- function(anno_type, anno, peaks=NULL, coverage=NULL, 
                            n_bg_seqs, median_peak_length, bg_widths = NULL,
                            type = "bed", genome_seq = genome, 
                            prefix = "", suffix = "", outdir){
  if ( !is.null(peaks) ){  # all regions of the same type without peaks
    bg_region <- subsetByOverlaps(anno, peaks, invert = TRUE)
  } else if( !is.null(coverage) ){  # all regions without any reads
    bg_region <- anno[which(sum(coverage[anno]) == 0)]
  }
  # remove mitochondrial genome, patches and haplotypes
  bg_region <- bg_region[seqnames(bg_region) %in% c("X", "Y", as.character(1:19))]
  seqlevels(bg_region) <- seqlevelsInUse(bg_region)  # drop unused levels
  
  # discard all that are shorter than the longest peak
  if (is.null(bg_widths)) {
    if (!is.null(median_peak_length)){
      peak_lengths <- median_peak_length
      bg_region <- bg_region[width(bg_region) >= peak_lengths]
      
    } else {
      stop(paste0("The peak length is unknown, please supply one of 
                  'median_peak_length' or 'bg_widths'"))
    }
  } else {
    peak_lengths <- bg_widths
    bg_region <- bg_region[width(bg_region) >= max(peak_lengths)]
  }

  # randomly choose regions and start positions for the bg seqs
  bg_region <- sample(bg_region, n_bg_seqs, replace = TRUE)
  starts <- floor(start(bg_region) + 
                    (runif(n_bg_seqs) * (width(bg_region) - peak_lengths + 1)))
  # starts <- sapply(bg_region, function(s) 
  #   sample(start(x):(end(x)-median_peak_length), 1))
  
  bg_gr <- GRanges(seqnames = seqnames(bg_region), 
                   ranges = IRanges(start = starts, 
                                    end = starts + peak_lengths - 1),
                   strand = strand(bg_region))
  seqlevelsStyle(bg_gr) <- "UCSC"
  
  if (type == "fasta"){
    bg_seq <- getSeq(genome_seq, bg_gr)
    export(bg_seq, 
           con = file.path(outdir, paste0(prefix, "_bg_", anno_type, "_", 
                                          n_bg_seqs, suffix, ".fasta")))
  } else if (type == "bed"){
    export(bg_gr, 
           con = file.path(outdir, paste0(prefix, "_bg_", anno_type, "_", 
                                          n_bg_seqs,suffix, ".bed")))
  } else if (type == "both"){
    bg_seq <- getSeq(genome_seq, bg_gr)
    export(bg_seq, 
           con = file.path(outdir, paste0(prefix, "_bg_", anno_type, "_", 
                                          n_bg_seqs, suffix, ".fasta")))
    export(bg_gr, 
           con = file.path(outdir, paste0(prefix, "_bg_", anno_type, "_", 
                                          n_bg_seqs,suffix, ".bed")))
  } else {
    print(paste0("type ", type, " unknown! Please pick one of fasta or bed."))
  }
}
```

We need to CLIP-seq coverage to make sure that we are not simmulating reads from bound regions.

```{r read-coverage, cache = TRUE}
cov <- list()
for (sample in c("HOMO_70K", "SNS_70K")){
  ga <- readGAlignmentPairs(here("BAM_deduplicated", sample, paste0(sample, "_deduplicated.bam")))
  cov[[sample]] <- coverage(ga)
  rm(ga)
  gc()
}
```


We generate background sequences that match the length distribution of the selected peaks.

```{r generate-bg-sns, cache = TRUE}
for (a in names(anno)){
  cat(a)
  export(getSeq(genome, sns_anno[[a]]), 
         con = file.path(here("analysis/deduplicated/MA_plot_selection/peaks_fasta"),
                         paste0("SNS_selected_peaks_", length(sns_anno[[a]]),
                                "_peaks_", a, ".fasta")), 
            format = "fasta")
  ws <- width(sns_anno[[a]])
  bg_ws <- sample(ws, size = 200000, replace = TRUE)

  generate_bg_seq(a, anno[[a]], coverage = cov[["SNS_70K"]], n_bg_seqs = 200000, 
                  bg_widths = bg_ws, type="fasta", prefix = "SNS", 
                  suffix = "_0reads_length_matched", 
                  outdir = here("analysis/deduplicated/MA_plot_selection/peaks_fasta"))
}
```

```{r generate-bg-cortex, cache = TRUE}
for (a in names(anno)){
  cat(a)
  export(getSeq(genome, hom_anno[[a]]), 
         con = file.path(here("analysis/deduplicated/MA_plot_selection/peaks_fasta"),
                         paste0("total_cortex_selected_peaks_", length(hom_anno[[a]]),
                                "_peaks_", a, ".fasta")), 
            format = "fasta")
  ws <- width(hom_anno[[a]])
  bg_ws <- sample(ws, size = 200000, replace = TRUE)

  generate_bg_seq(a, anno[[a]], coverage = cov[["HOMO_70K"]], n_bg_seqs = 200000, 
                  bg_widths = bg_ws, type="fasta", prefix = "total_cortex", 
                  suffix = "_0reads_length_matched" ,
                  outdir = here("analysis/deduplicated/MA_plot_selection/peaks_fasta"))
}
```


# Centered window on peaks

We center a window of size 41 on the peak region with the highest coverage. This should improve the motif finding because all sequences will have the same lengths.

```{r QC-peak-center, cache = TRUE}
## how many peaks have more than one position with the max # reads? 
## --> most peaks have only few positions, max is 50
sns <- import(here("analysis/deduplicated/MA_plot_selection/selected_peaks/selected_peaks_SNS_pValue_1e-05.bed"))
hom <- import(here("analysis/deduplicated/MA_plot_selection/selected_peaks/selected_peaks_total_cortex_pValue_3e-03.bed"))
max_peak_lengths <- sapply(which(cov[["SNS_70K"]][sns] == max(cov[["SNS_70K"]][sns])), length)
table(max_peak_lengths)

## If a peak has more than one max position, we take all of them and center the
## window around it. So we need to make sure that the window covers the region with max coverage
## Where are the positions with max coverage?
which(cov[["SNS_70K"]][sns] == max(cov[["SNS_70K"]][sns]))

# How far are the positions apart?
## range of positions with max value
a <- range(which(cov[["SNS_70K"]][sns] == max(cov[["SNS_70K"]][sns])))
table(a[,2] - a[,1]) ## difference of positions
## --> they are no further than 51 nt apart
## --> window of length 41 convers most peaks

## we take the median of all positions with the max coverage as the center
sns$peak_center <- start(sns) + 
  median(which(cov[["SNS_70K"]][sns] == max(cov[["SNS_70K"]][sns]))) -1

hom$peak_center <- start(hom) + 
  median(which(cov[["HOMO_70K"]][hom] == max(cov[["HOMO_70K"]][hom]))) -1
```

<!-- We annotate the peaks. -->
<!-- ```{r peak-center-annotation, cache = TRUE} -->
<!-- sns_anno <- lapply(anno, function(x) { -->
<!--   res <- subsetByOverlaps(sns, x, type = "any") -->
<!--   seqlevelsStyle(res) <- "UCSC" -->
<!--   res}) -->

<!-- hom_anno <- lapply(anno, function(x) { -->
<!--   res <- subsetByOverlaps(hom, x, type = "any") -->
<!--   seqlevelsStyle(res) <- "UCSC" -->
<!--   res}) -->
<!-- names(hom_anno) <- names(anno) -->
<!-- ``` -->

We export the window sequences and generate background sequences of the same length.
```{r generate-seq-peak-window, cache = TRUE}
write_bed_fasta <- function(win_size = 20, peaks, sample = "SNS", anno, cov){
  ## Write BED of peak windows
  win <- peaks
  start(win) <- win$peak_center - win_size
  end(win) <- win$peak_center + win_size
  export(win, 
         con = file.path(here("analysis/deduplicated/MA_plot_selection/selected_peaks"),
                         paste0(sample, "_selected_peaks_", length(win), 
                                "_peaks_window", 2 * win_size + 1, ".bed")), 
          format = "bed")
  seqlevelsStyle(win) <- "UCSC"
  export(getSeq(genome, win), 
         con = file.path(here("analysis/deduplicated/MA_plot_selection/peaks_fasta"),
                         paste0(sample, "_selected_peaks_", length(win),
                                "_peaks_window", 2 * win_size + 1, ".fasta")),
         format = "fasta") 
  
  peaks_anno <- lapply(anno, function(x) {
  res <- subsetByOverlaps(peaks, x, type = "any")
  seqlevelsStyle(res) <- "UCSC"
  res})
  
  ## Write FASTA for each location
  for (a in names(anno)){
    cat(a)
    ## Write BED file with the peak windows
    win <- peaks_anno[[a]]
    start(win) <- win$peak_center - win_size
    end(win) <- win$peak_center + win_size
  
    export(getSeq(genome, win), 
           con = file.path(here("analysis/deduplicated/MA_plot_selection/peaks_fasta"),
                           paste0(sample, "_selected_peaks_", length(win),
                                  "_peaks_", a, "_window", 2 * win_size + 1, ".fasta")), 
              format = "fasta")
    generate_bg_seq(a, anno[[a]], coverage = cov, n_bg_seqs = 200000, 
                    median_peak_length = win_size * 2 + 1, type="fasta", 
                    prefix = sample, 
                    suffix = paste0("_0reads_window", win_size * 2 + 1), 
                    outdir = here("analysis/deduplicated/MA_plot_selection/peaks_fasta"))
  }
}

write_bed_fasta(win_size = 20, sns, sample = "SNS", anno, cov = cov[["SNS_70K"]])
write_bed_fasta(win_size = 20, hom, sample = "total_cortex", anno, cov = cov[["HOMO_70K"]])
```

Generate background sequences for all peaks, not seperated by location.

```{r generate-bg-peak-window-all}
win_size <- 20
generate_bg_seq("all_peaks", unique(unlist(anno)), coverage = cov[["SNS_70K"]], 
                n_bg_seqs = 200000, median_peak_length = win_size * 2 + 1, 
                type="fasta", prefix = "SNS", 
                suffix = paste0("_0reads_window", win_size * 2 + 1), 
                outdir = here("analysis/deduplicated/MA_plot_selection/peaks_fasta"))

generate_bg_seq("all_peaks", unique(unlist(anno)), coverage = cov[["HOMO_70K"]], 
                n_bg_seqs = 200000, median_peak_length = win_size * 2 + 1, 
                type="fasta", prefix = "total_cortex", 
                suffix = paste0("_0reads_window", win_size * 2 + 1),
                outdir = here("analysis/deduplicated/MA_plot_selection/peaks_fasta"))
```

