---
title: "omniCLIP peak filtering"
author: "Katharina Hembach"
date: "5/28/2019"
output: 
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup

```{r packages, warning = FALSE, message = FALSE}
library(here)
library(rtracklayer)
library(GenomicFeatures)
library(stringr)
library(ggplot2)
library(tidyr)
```

# Reading omniCLIP peaks

We read the top peaks and and filter the top peaks based on score.

```{r omniCLIP, cache = 2}
## here sets the root directory to the github directory at
## "/home/Shared/data/seq/sonu_CLIPseq/clip_March2018"

gtf_file <- "/home/Shared/data/annotation/Mouse/Ensembl_GRCm38.90/gtf/Mus_musculus.GRCm38.90.gtf"
out_dir <- here("analysis", "omniCLIP", "scoreFix")

metadat <- read.table(here("metadata.txt"), header = TRUE)
gtf <- import(gtf_file)
cutoff <- 100
hom_score <- 30

###### Read the CLIP peaks -----------------------------------------------------
omni <- list()
for (i in c(1, 3)) {
  sample <- as.character(metadat$ID[i])
  omni[[sample]] <- import(here("omniCLIP", paste0(sample, "_scoreFix"), 
                                paste0(sample,"_bg_", metadat$group[i], 
                                       "_scoreFix_pred.bed")))
  omni[[sample]]$gene_id <- str_split(omni[[sample]]$name, "_", simplify = TRUE)[,1]
}
## number of peaks
lengths(omni)


### filter top peaks
omni_filtered <- list()
for (sample in c("HOMO_70K", "SNS_70K")){
  omni_filtered[[sample]] <- omni[[sample]][omni[[sample]]$score >= cutoff]
}
lengths(omni_filtered)
```

# SNS peak filtering

Most of the SNS peaks are located in introns. However, preRNA should not be present in the synapse. We think that intronic peaks are caused by contamination or the polyA RNA-seq background that we use for omniCLIP. RNA-seq in general, has less reads in intronic regions than in exons of UTRs, Therefore, intronic CLIP-seq peaks might get better scores than exonic peaks simply because the bg is missing reads in the intronic peak region but not in the exon.
To remove intronic peaks that might be caused by contamination, we overlap the SNS peaks with the homogenate peaks (with a score above a certain threshold). We remove all SNS peaks where the homogenate samples also has a peak in the same region. This way, we might be removign true peaks, but getting rid of contamination is more important to us.

We take all homogenate peaks that have a score of >= `r hom_score` to filter the SNS peaks.


```{r sns_homogenate_overlaps, cache = 2}
hom_bg <- omni[["HOMO_70K"]][omni[["HOMO_70K"]]$score >= hom_score]
## Filter criteria: at least 50% of the shortes peak overlaps with a homogenate peak
min_overlap <- ceiling(min(width(omni_filtered[["SNS_70K"]]))/2)
filtered <- subsetByOverlaps(omni_filtered[["SNS_70K"]], hom_bg, type = "any", 
                             invert = TRUE, minoverlap = min_overlap)
length(filtered)      

export(filtered, here("analysis", "omniCLIP", "scoreFix", "top_100", 
                      paste0("SNS_70K_cutoff", cutoff, 
                             "_filtered_homogenate_bg_cutoff", hom_score, 
                             ".bed")))

omni_list <- omni_filtered
omni_list[["SNS_filtered"]] <- filtered
```

# Genes of the filterd peaks

```{r gene_list, cache = 2}
## Write table with the gene_id, the gene name and the number of peaks in each gene
add_gene_annotation <- function(peaks, genes){
  m <- match(peaks$gene_id, genes$gene_id)
  res <- peaks[, "score"]
  mcols(res) <- cbind(mcols(res), mcols(genes[m, c("gene_id", "gene_name", 
                                                   "gene_biotype")]))
  res
}

filtered <- add_gene_annotation(filtered, anno[["gene"]])

res <- filtered %>% 
  as.data.frame() %>% 
  dplyr::group_by(gene_id) %>%
  dplyr::select(seqnames, gene_id, gene_name, gene_biotype) %>%
  dplyr::mutate(nr_peaks = n()) %>%
  dplyr::arrange(desc(nr_peaks)) %>%
  unique()
print(res, n = 40)
```


# Location of the filtered peaks

Where in a gene are the filtered peaks located?
 
```{r prepare_annotation, cache = 2, message = FALSE, warning = FALSE}
## annotation: seperate into exon, intron, 3'UTR, 5'UTR or exon + 3'UTR/5'UTR 
## in case the exons are overlapping
exon <- gtf[gtf$type == "exon"] %>% unique
gene <- gtf[gtf$type == "gene"] %>% unique
five_utr <- gtf[gtf$type == "five_prime_utr"] %>% unique
three_utr <- gtf[gtf$type == "three_prime_utr"] %>% unique

exon_three_utr <- exon[overlapsAny(exon, three_utr, type = "any")]
exon_five_utr <- exon[overlapsAny(exon, five_utr, type = "any")]
exon_unique <- exon[!overlapsAny(exon, c(five_utr, three_utr),  type = "any")]
exon_three_utr <- c(exon_three_utr, three_utr) %>% unique
exon_five_utr <- c(exon_five_utr, five_utr) %>% unique

## the set of annotations is the following, because all 3' and 5' UTR regions
## are overlapped by an exon
anno <- list(gene = gene, exon = exon_unique, three_prime_utr = exon_three_utr, 
             five_prime_utr = exon_five_utr)

## intron annotation
txdb <- makeTxDbFromGRanges(gtf)
introns <- unlist(intronsByTranscript(txdb))
## remove the intronic parts that overlap with exons from other transcripts
anno[["intron"]] <- setdiff(introns, anno[["exon"]])
```

```{r gene_region, cache = 2, message = FALSE}

peak_location_barplot <- function(peaks) {
  omni_olap_len <- list()
  for (sample in names(peaks)){
    omni_olap_len[[sample]] <- lapply(anno, function(x) 
      length(subsetByOverlaps(peaks[[sample]], x, type = "any")))
    names(omni_olap_len[[sample]]) <- names(anno)
  } 
  df <- as.data.frame(t(cbind(sapply(omni_olap_len, as.data.frame))))
  
  df$peaks <- rownames(df)
  df[,names(df) != "peaks"] <- apply(df[,names(df) != "peaks"], 2, as.integer)
  
  df <- df %>% gather(key = "annotation", value = "peak_number", -peaks)
  df <- df[df$annotation != "gene",]
  df$percentage <- df$peak_number / sapply(df$peaks, function(x) 
    sum(df[df$peaks == x, "peak_number"]) ) * 100
  
  ## reorder the annotation factor levels
  df$annotation <- factor(df$annotation, 
                          levels = c("exon", "five_prime_utr", 
                                     "three_prime_utr", "intron"))
  
  ## stacked barplot with the percentage of reads in the different gene regions
  p <- ggplot(df, aes(x = peaks, y = percentage, fill = annotation)) +
    geom_col(position = "stack") +
    theme_bw() +
    theme(text=element_text(size=25), axis.text.y = element_text(angle = 45, 
                                                                 hjust = 1) ) +
    coord_flip() +
    theme(legend.position="bottom", legend.direction="vertical") 
  p
}

## all peaks
peak_location_barplot(omni_list)
```
